core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java

```diff
@@ -0,0 +1,292 @@
+   * the stateChangeLock is already acquired in the caller before calling this function.
+   */
+  private int readInternal(byte[] b, int offset, int len) throws IOException {
+    assert (stateChangeLock.isLocked());
+    if (offset < 0 || len < 0 || offset + len < 0 || offset + len > b.length) {
+      throw new IndexOutOfBoundsException();
+    }
+    if (!activeBuffer.hasRemaining()) {
+      if (!isReadInProgress) {
+        // This condition will only be triggered for the first time read is called.
+        readAsync(activeBuffer);
+      }
+      waitForAsyncReadComplete();
+    }
+
+    if (isReadAborted) {
+      throw new IOException(readException);
+    }
+    if (isEndOfStream()) {
+      return -1;
+    }
+    len = Math.min(len, activeBuffer.remaining());
+    activeBuffer.get(b, offset, len);
+
+    if (activeBuffer.remaining() <= readAheadThresholdInBytes && !readAheadBuffer.hasRemaining()) {
+      readAsync(readAheadBuffer);
+    }
+    if (!activeBuffer.hasRemaining()) {
+      ByteBuffer temp = activeBuffer;
+      activeBuffer = readAheadBuffer;
+      readAheadBuffer = temp;
+    }
+    return len;
+  }
+
+  @Override
+  public synchronized int available() throws IOException {
+    stateChangeLock.lock();
+    // Make sure we have no integer overflow.
+    int val = (int) Math.min((long) Integer.MAX_VALUE,
+            (long) activeBuffer.remaining() + readAheadBuffer.remaining());
+    stateChangeLock.unlock();
+    return val;
+  }
+
+  @Override
+  public synchronized long skip(long n) throws IOException {
+    stateChangeLock.lock();
+    long skipped;
+    try {
+      skipped = skipInternal(n);
+    } finally {
+      stateChangeLock.unlock();
+    }
+    return skipped;
+  }
+
+  /**
+   * Internal skip function which should be called only from skip() api. The assumption is that
+   * the stateChangeLock is already acquired in the caller before calling this function.
+   */
+  private long skipInternal(long n) throws IOException {
+    assert (stateChangeLock.isLocked());
+    if (n <= 0L) {
+      return 0L;
+    }
+    if (isReadInProgress) {
+      waitForAsyncReadComplete();
+    }
+    if (available() >= n) {
+      // we can skip from the internal buffers
+      int toSkip = (int)n;
+      byte[] temp = new byte[toSkip];
+      while (toSkip > 0) {
+        int skippedBytes = read(temp, 0, toSkip);
+        toSkip -= skippedBytes;
+      }
+      return n;
+    }
+    int skippedBytes = available();
+    long toSkip = n - skippedBytes;
+    activeBuffer.position(0);
+    activeBuffer.flip();
+    readAheadBuffer.position(0);
+    readAheadBuffer.flip();
+    long skippedFromInputStream = underlyingInputStream.skip(toSkip);
+    return skippedBytes + skippedFromInputStream;
+  }
+
+  @Override
+  public synchronized void close() throws IOException {
+    executorService.shutdown();
+    StorageUtils.dispose(activeBuffer);
+    StorageUtils.dispose(readAheadBuffer);
```

Comment:
await termination of `executorService` ?
ReadAheadInputStream can continue to read after close is invoked : which will be happening on stream which might be closed and buffers which might be disposed.

Access to activeBuffer/readAheadBuffer should be done within `stateChangeLock` ? 
Also, delegate `close()` to underlyingInputStream ?

Reply:
Good point, fixed it.